src/main/java/edu/harvard/iq/dataverse/engine/command/impl/CuratePublishedDatasetVersionCommand.java

```diff
@@ -85,18 +85,26 @@ public Dataset execute(CommandContext ctxt) throws CommandException {
         Dataset tempDataset = ctxt.em().merge(getDataset());
 
         // Look for file metadata changes and update published metadata if needed
-        for (DataFile dataFile : tempDataset.getFiles()) {
-            List<FileMetadata> fmdList = dataFile.getFileMetadatas();
+        List<FileMetadata> pubFmds = updateVersion.getFileMetadatas();
+        int pubFileCount = pubFmds.size();
+        int newFileCount = tempDataset.getEditVersion().getFileMetadatas().size();
+        if (pubFileCount != newFileCount) {
+            logger.severe("Draft version of dataset: " + tempDataset.getId() + " has: " + newFileCount + " while last published version has " + pubFileCount);
+            throw new IllegalCommandException("Different number of files in draft version", this);
+        }
+        for (FileMetadata publishedFmd : pubFmds) {
+            DataFile dataFile = publishedFmd.getDataFile();
             FileMetadata draftFmd = dataFile.getLatestFileMetadata();
-            FileMetadata publishedFmd = null;
-            for (FileMetadata fmd : fmdList) {
-                if (fmd.getDatasetVersion().equals(updateVersion)) {
-                    publishedFmd = fmd;
-                    break;
-                }
-            }
             boolean metadataUpdated = false;
-            if (draftFmd != null && publishedFmd != null) {
+            if (draftFmd == null || draftFmd.getDatasetVersion().equals(updateVersion)) {
+                if (draftFmd == null) {
+                    logger.severe("Unable to find latest FMD for file id: " + dataFile.getId());
+                } else {
+                    logger.severe("No filemetadata for file id: " + dataFile.getId() + " in draft version");
+                }
+                throw new IllegalCommandException("Cannot change files in the dataset", this);
```

Comment:
English should go in the bundle. In addition, since we have the file id, maybe it would be worth exposing it to the user in the error message, if it will help them troubleshoot the problem.